src/utils/DragAndDrop.js

```diff
@@ -155,39 +147,94 @@ define(function (require, exports, module) {
-            
-            var files = event.dataTransfer.files;
-            if (files && files.length) {
-                event.stopPropagation();
-                event.preventDefault();
-                
-                brackets.app.getDroppedFiles(function (err, paths) {
-                    if (!err) {
-                        openDroppedFiles(paths);
+            event.stopPropagation();
+            event.preventDefault();
+
+            var pathList = [];
+            var errorList = [];
+
+            function prepareDropPaths(fileList) {
+                // Convert FileList object to an Array with all *.html files at the end,
+                // since we need to write any .css, .js, etc. resources first such that
+                // Blob URLs can be generated for these resources prior to rewriting an HTML file.
+                return _.toArray(fileList).sort(function(a,b) {
+                    a = Content.isHTML(Path.extname(a.name)) ? 10 : 1;
+                    b = Content.isHTML(Path.extname(b.name)) ? 10 : 1;
+
+                    if(a < b) {
+                        return -1;
+                    }
+                    if(a > b) {
+                        return 1;
                     }
+                    return 0;
                 });
             }
+
+            function importFile(item) {
+                var deferred = new $.Deferred();
+                var reader = new FileReader();
+
+                reader.onload = function(e) {
+                    delete reader.onload;
+
+                    var filename = Path.join('/', item.name);
+                    var file = FileSystem.getFileForPath(filename);
+
+                    // Create a Filer Buffer, and determine the proper encoding
+                    var buffer = new Filer.Buffer(e.target.result);
+                    var encoding = Content.isUTF8Encoded(Path.extname(filename)) ? 'utf8' : null;
+                    if(encoding === 'utf8') {
+                        buffer = buffer.toString();
+                    }
+
+                    file.write(buffer, {encoding: encoding}, function(err) {
+                        if (err) {
+                            deferred.reject(err);
+                            return;
+                        }
+
+                        pathList.push(filename);
+                        deferred.resolve();
+                    });
+                };
+
+                // Deal with error cases, for example, trying to drop a folder vs. file
+                reader.onerror = function(err) {
+                    delete reader.onerror;
+
+                    errorList.push({path: item.name, error: "Unable to read dropped item."});
+                    deferred.reject(err);
+                };
+
+                reader.readAsArrayBuffer(item);
+
+                return deferred.promise();
+            }
+
+            var files = event.dataTransfer.files;
+
+            if (files && files.length) {
```

Comment:
Above you mention that `prepareDropPaths` converts the object passed into an array, so is `files` an object? And will it have the `length` property?